Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+{
+    char buf[128];
+    char *msg;
+    assert(ZSTD_isError(zstd_ret));
+
+    switch (type)
+    {
+    case ERR_DECOMPRESS:
+        msg = "Unable to decompress zstd data: %s";
+        break;
+    case ERR_COMPRESS:
+        msg = "Unable to compress zstd data: %s";
+        break;
+    case ERR_SET_PLEDGED_INPUT_SIZE:
+        msg = "Unable to set pledged uncompressed content size: %s";
+        break;
+
+    case ERR_LOAD_D_DICT:
+        msg = "Unable to load zstd dictionary or prefix for decompression: %s";
+        break;
+    case ERR_LOAD_C_DICT:
+        msg = "Unable to load zstd dictionary or prefix for compression: %s";
+        break;
+
+    case ERR_GET_C_BOUNDS:
+        msg = "Unable to get zstd compression parameter bounds: %s";
+        break;
+    case ERR_GET_D_BOUNDS:
+        msg = "Unable to get zstd decompression parameter bounds: %s";
+        break;
+    case ERR_SET_C_LEVEL:
+        msg = "Unable to set zstd compression level: %s";
+        break;
+
+    case ERR_TRAIN_DICT:
+        msg = "Unable to train zstd dictionary: %s";
+        break;
+    case ERR_FINALIZE_DICT:
+        msg = "Unable to finalize zstd dictionary: %s";
+        break;
+
+    default:
+        Py_UNREACHABLE();
+    }
+    PyOS_snprintf(buf, sizeof(buf), msg, ZSTD_getErrorName(zstd_ret));
+    PyErr_SetString(state->ZstdError, buf);
+}
+
+typedef struct {
+    const int parameter;
+    const char parameter_name[32];
+} ParameterInfo;
+
+static const ParameterInfo cp_list[] =
+{
+    {ZSTD_c_compressionLevel, "compressionLevel"},
+    {ZSTD_c_windowLog,        "windowLog"},
+    {ZSTD_c_hashLog,          "hashLog"},
+    {ZSTD_c_chainLog,         "chainLog"},
+    {ZSTD_c_searchLog,        "searchLog"},
+    {ZSTD_c_minMatch,         "minMatch"},
+    {ZSTD_c_targetLength,     "targetLength"},
+    {ZSTD_c_strategy,         "strategy"},
+
+    {ZSTD_c_enableLongDistanceMatching, "enableLongDistanceMatching"},
+    {ZSTD_c_ldmHashLog,       "ldmHashLog"},
+    {ZSTD_c_ldmMinMatch,      "ldmMinMatch"},
+    {ZSTD_c_ldmBucketSizeLog, "ldmBucketSizeLog"},
+    {ZSTD_c_ldmHashRateLog,   "ldmHashRateLog"},
+
+    {ZSTD_c_contentSizeFlag,  "contentSizeFlag"},
+    {ZSTD_c_checksumFlag,     "checksumFlag"},
+    {ZSTD_c_dictIDFlag,       "dictIDFlag"},
+
+    {ZSTD_c_nbWorkers,        "nbWorkers"},
+    {ZSTD_c_jobSize,          "jobSize"},
+    {ZSTD_c_overlapLog,       "overlapLog"}
+};
+
+static const ParameterInfo dp_list[] =
+{
+    {ZSTD_d_windowLogMax, "windowLogMax"}
+};
+
+void
+set_parameter_error(const _zstd_state* const state, int is_compress,
+                    int key_v, int value_v)
+{
+    ParameterInfo const *list;
+    int list_size;
+    char const *name;
+    char *type;
+    ZSTD_bounds bounds;
+    int i;
+    char pos_msg[128];
+
+    if (is_compress) {
+        list = cp_list;
+        list_size = Py_ARRAY_LENGTH(cp_list);
+        type = "compression";
+    } else {
```

Comment:
Nit, but can you (in a follow-up PR if you want to) make it so that the new code follow PEP-7? this includes writing something like:

```c
if (...) {
...
}
else {
...
}
```

Reply:
I plan on making the code PEP 7 compliant prior to merge, but want to work through the semantic/functional issues first.

Reply:
I've fixed this but considering the timing I'll work on a more thorough review of style to make the code PEP 7 compliant once everything is landed.